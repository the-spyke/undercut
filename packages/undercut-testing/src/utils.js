export function getArrayObserver(array) {
	return {
		next(value) {
			array.push(value);
		},
		return() {
			// Empty.
		},
		throw(error) {
			throw error;
		},
	};
}

export function abort(coroutine, error) {
	if (coroutine.throw) {
		coroutine.throw(error);
	}

	throw error;
}

export function asObserver(generator) {
	return function (...args) {
		const observer = generator(...args);

		observer.next();

		return observer;
	};
}

export function close(coroutine) {
	if (coroutine.return) {
		coroutine.return();
	}
}
