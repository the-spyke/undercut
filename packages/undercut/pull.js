import * as compare from "./src/utils/compare.js";

export * from "./src/pull/pull_core.js";
export * from "./src/pull/pull_operations.js";
export * from "./src/pull/pull_sources.js";
export * from "./src/pull/pull_targets.js";

export { compare };
export * from "./src/utils/coroutine.js";
export * from "./src/utils/iterable.js";
export { asc, desc } from "./src/utils/ordering.js";
