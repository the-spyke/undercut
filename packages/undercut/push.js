export * from "./src/push/push_core.js";
export * from "./src/push/push_operations.js";
export * from "./src/push/push_targets.js";
