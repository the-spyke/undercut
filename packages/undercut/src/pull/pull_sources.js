import { assert } from "../utils/assert.js";
import { createIterable } from "../utils/iterable.js";

function* rangeGen(start, end, step) {
	if (start < end) {
		for (let i = start; i < end; i += step) {
			yield i;
		}
	} else if (start > end) {
		for (let i = start; i > end; i -= step) {
			yield i;
		}
	}
}

/**
 * Creates an iterable of numbers in range `[start, end)` with optional step.
 * @param {Number} start
 * @param {Number} end
 * @param {Number} [step=1]
 * @returns {Iterable<Number>}
 */
export function range(start, end, step = 1) {
	assert(Number.isFinite(start), `"start" is required, must be a number.`);
	assert(Number.isFinite(end), `"end" is required, must be a number.`);
	assert(Number.isFinite(step) && step > 0, `"step" must be an absolute non-zero number.`);

	return createIterable(() => rangeGen(start, end, step));
}
