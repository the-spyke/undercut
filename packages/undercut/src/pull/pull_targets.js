import { assert } from "../utils/assert.js";
import { abort, close } from "../utils/coroutine.js";
import { isObserver } from "../utils/language.js";

/**
 * @param {Iterable} iterable
 * @returns {Array}
 */
export function toArray(iterable) {
	return Array.from(iterable);
}

/**
 * @param {Function} consumer
 * @returns {Function}
 */
export function toConsumer(consumer) {
	return function (iterable) {
		let index = 0;

		for (const item of iterable) {
			consumer(item, index);
			index++;
		}
	};
}

/**
 * @param {Iterable} iterable
 * @returns {Map}
 */
export function toMap(iterable) {
	return new Map(iterable);
}

/**
 * @param {Iterable} iterable
 * @returns {void}
 */
export function toNull(iterable) {
	for (const item of iterable) { // eslint-disable-line no-unused-vars
		// Do nothing.
	}
}

/**
 * @param {Iterable} iterable
 * @returns {Object}
 */
export function toObject(iterable) {
	return Object.fromEntries(iterable);
}

/**
 * @param {Observer} observer
 * @returns {Function}
 */
export function toPushLine(observer) {
	assert(isObserver(observer), `"observer" is required and must be an Observable.`);

	return function (iterable) {
		try {
			for (const item of iterable) {
				observer.next(item);
			}
		} catch (error) {
			abort(observer, error);
		} finally {
			close(observer);
		}

		return observer;
	};
}

/**
 * @param {Iterable} iterable
 * @returns {Set}
 */
export function toSet(iterable) {
	return new Set(iterable);
}

/**
 * @param {Iterable} iterable
 * @returns {any}
 */
export function toValue(iterable) {
	let firstValue = undefined;
	let count = 0;

	for (const item of iterable) {
		assert(count === 0, `"toValue()" may be applied only to a sequence of one item, but got at least 2.`);

		firstValue = item;
		count++;
	}

	assert(count === 1, `"toValue()" may be applied only to a sequence of one item, but got none.`);

	return firstValue;
}
