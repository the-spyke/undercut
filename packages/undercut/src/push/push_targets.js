import { assert, assertFunctor } from "../utils/assert.js";
import { asObserver } from "../utils/coroutine.js";
import { rethrow } from "../utils/function.js";
import { isUndefined, isFunction } from "../utils/language.js";

export function toArray() {
	return {
		next(value) {
			this.values.push(value);
		},
		return() {
			// Empty.
		},
		throw: rethrow,
		values: [],
	};
}

const consumerTarget = asObserver(function* (consumer, finalizer) {
	assertFunctor(consumer, `consumer`);
	assert(isUndefined(finalizer) || isFunction(finalizer), `"finalizer" is optional, could be a function.`);

	let error = undefined;
	let index = 0;

	try {
		while (true) {
			consumer(yield, index);
			index++;
		}
	} catch (e) {
		error = e;
		throw e;
	} finally {
		if (finalizer) {
			finalizer(error, index);
		}
	}
});

/**
 * @param {Function} consumer
 * @param {Function} [finalizer]
 * @returns {Observer}
 */
export function toConsumer(consumer, finalizer) {
	return consumerTarget(consumer, finalizer);
}

/**
 * @returns {Observer}
 */
export function toNull() {
	return {
		next() {
			// Empty.
		},
		return() {
			// Empty.
		},
		throw(e) {
			throw e;
		}
	};
}

const valueTarget = asObserver(function* (setter) {
	assertFunctor(setter, `setter`);

	let firstValue = undefined;
	let count = 0;

	try {
		firstValue = yield;
		count++;
		yield;
		count++;

		assert(count === 1, `"toValue()" may be applied only to a sequence of one item, but got at least 2.`);
	} finally {
		assert(count > 0, `"toValue()" may be applied only to a sequence of one item, but got none.`);

		if (count === 1) {
			setter(firstValue);
		}
	}
});

/**
 * @param {Function} setter
 * @returns {Observer}
 */
export function toValue(setter) {
	return valueTarget(setter);
}
