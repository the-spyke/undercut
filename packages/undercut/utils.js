import * as compare from "./src/utils/compare.js";

export { compare };
export * from "./src/utils/function.js";
export * from "./src/utils/language.js";
export { asc, desc } from "./src/utils/ordering.js";
export * from "./src/utils/promise.js";
