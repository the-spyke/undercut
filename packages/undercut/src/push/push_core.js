import { assert, assertPipeline, assertSource } from "../utils/assert.js";
import { isObserver, isFunction } from "../utils/language.js";
import { closeObserver, initializeObserver } from "../utils/observer.js";

import { toArray } from "./push_targets.js";

const operationErrorMessage = `An operation must be a function taking and returning an Observer.`;

function connectPipeline(pipeline, target, initialize = true) {
	assertPipeline(pipeline);
	assert(isObserver(target), `"target" is required and must be an Observer.`);

	const operations = Array.isArray(pipeline) ? pipeline : [...pipeline];

	let observer = target;

	for (let index = operations.length - 1; index >= 0; index--) {
		const operation = operations[index];

		assert(isFunction(operation), operationErrorMessage);

		observer = operation(observer);

		assert(isObserver(observer), operationErrorMessage);

		if (index > 0 || initialize) {
			initializeObserver(observer);
		}
	}

	return observer;
}

export function composeOperations(operations) {
	return function (observer) {
		return connectPipeline(operations, observer, false);
	};
}

export function pushLine(pipeline, target) {
	return connectPipeline(pipeline, target, true);
}

export function push(target, pipeline, source) {
	assertSource(source);

	const observer = pushLine(pipeline, target);

	try {
		for (const item of source) {
			observer.next(item);
		}
	} finally {
		closeObserver(observer);
	}

	return target;
}

export function pushItems(pipeline, source) {
	const target = toArray();

	push(target, pipeline, source);

	return target.values;
}
