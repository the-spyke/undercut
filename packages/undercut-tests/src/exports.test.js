import * as pullExports from "@undercut/pull";
import * as pushExports from "@undercut/push";
import * as testingExports from "@undercut/testing";
import * as utilsExports from "@undercut/utils";

describe(`exports`, () => {
	test(`@undercut/pull`, () => {
		expect(Object.keys(pullExports).sort()).toMatchSnapshot();
	});

	test(`@undercut/push`, () => {
		expect(Object.keys(pushExports).sort()).toMatchSnapshot();
	});

	test(`@undercut/testing`, () => {
		expect(Object.keys(testingExports).sort()).toMatchSnapshot();
	});

	test(`@undercut/utils`, () => {
		expect(Object.keys(utilsExports).sort()).toMatchSnapshot();
	});
});
