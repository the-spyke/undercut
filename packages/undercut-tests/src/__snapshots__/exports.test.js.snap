// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exports of @undercut/pull should be stable 1`] = `
Array [
  "Cohort",
  "abort",
  "append",
  "asObserver",
  "asUnclosable",
  "asc",
  "average",
  "buffer",
  "bufferAll",
  "chunk",
  "close",
  "collect",
  "collectArray",
  "collectMap",
  "collectObject",
  "collectSet",
  "compact",
  "compare",
  "composeOperations",
  "concatEnd",
  "concatStart",
  "count",
  "createIterable",
  "desc",
  "difference",
  "differenceBy",
  "every",
  "filter",
  "find",
  "findIndex",
  "first",
  "flatMap",
  "flatten",
  "flattenArrays",
  "forEach",
  "getIterator",
  "getRecursiveMapper",
  "groupBy",
  "includes",
  "interleave",
  "intersection",
  "intersectionBy",
  "join",
  "last",
  "map",
  "max",
  "min",
  "nth",
  "orderBy",
  "peekIterable",
  "prepend",
  "pull",
  "pullArray",
  "pullLine",
  "pullValue",
  "range",
  "reduce",
  "remove",
  "reverse",
  "skip",
  "skipWhile",
  "some",
  "sort",
  "sortNumbers",
  "sortStrings",
  "sum",
  "symmetricDifference",
  "symmetricDifferenceBy",
  "take",
  "takeWhile",
  "toArray",
  "toConsumer",
  "toMap",
  "toNull",
  "toObject",
  "toObserver",
  "toSet",
  "toValue",
  "union",
  "unionBy",
  "unique",
  "uniqueBy",
  "unzip",
  "unzipWith",
  "zip",
  "zipWith",
]
`;

exports[`Exports of @undercut/push should be stable 1`] = `
Array [
  "Cohort",
  "abort",
  "append",
  "asObserver",
  "asUnclosable",
  "asc",
  "average",
  "buffer",
  "bufferAll",
  "chunk",
  "close",
  "collect",
  "collectArray",
  "collectMap",
  "collectObject",
  "collectSet",
  "compact",
  "compare",
  "composeOperations",
  "concatEnd",
  "concatStart",
  "count",
  "createIterable",
  "desc",
  "difference",
  "differenceBy",
  "every",
  "filter",
  "find",
  "findIndex",
  "first",
  "flatMap",
  "flatten",
  "flattenArrays",
  "forEach",
  "getIterator",
  "getRecursiveMapper",
  "groupBy",
  "includes",
  "interleave",
  "intersection",
  "intersectionBy",
  "join",
  "last",
  "map",
  "max",
  "min",
  "nth",
  "orderBy",
  "peekIterable",
  "prepend",
  "push",
  "pushArray",
  "pushLine",
  "pushValue",
  "reduce",
  "remove",
  "reverse",
  "skip",
  "skipWhile",
  "some",
  "sort",
  "sortNumbers",
  "sortStrings",
  "sum",
  "symmetricDifference",
  "symmetricDifferenceBy",
  "take",
  "takeWhile",
  "toArray",
  "toConsumer",
  "toNull",
  "toValue",
  "union",
  "unionBy",
  "unique",
  "uniqueBy",
  "unzip",
  "unzipWith",
  "zip",
  "zipWith",
]
`;

exports[`Exports of @undercut/utils should be stable 1`] = `
Array [
  "Cohort",
  "abort",
  "asObserver",
  "asUnclosable",
  "asc",
  "close",
  "compare",
  "createIterable",
  "delay",
  "desc",
  "getDoneItem",
  "getIterator",
  "getRecursiveMapper",
  "hasOwnProps",
  "identity",
  "isArrayBuffer",
  "isBoolean",
  "isCoroutine",
  "isDate",
  "isDefined",
  "isError",
  "isFalsy",
  "isFunction",
  "isIterable",
  "isIterator",
  "isMap",
  "isNegative",
  "isNegativeOrZero",
  "isNull",
  "isNullish",
  "isNumber",
  "isNumberValue",
  "isObject",
  "isObjectValue",
  "isObserver",
  "isPlainObject",
  "isPositive",
  "isPositiveOrZero",
  "isRegExp",
  "isSet",
  "isString",
  "isSymbol",
  "isTruthy",
  "isUndefined",
  "isWeakMap",
  "isWeakSet",
  "negate",
  "noop",
  "not",
  "peekIterable",
  "rethrow",
  "unwrapPromise",
  "wait",
]
`;
